pkg/tso/keyspace_group_manager.go

```diff
@@ -148,6 +154,40 @@ func (s *state) getKeyspaceGroupMetaWithCheck(
 		mcsutils.DefaultKeyspaceGroupID, nil
 }
 
+func (s *state) getNextPrimaryToReset(
+	groupID int, localAddress string,
+) (member ElectionMember, kg *endpoint.KeyspaceGroup, localPriority, nextGroupID int) {
+	s.RLock()
+	defer s.RUnlock()
+
+	groupSize := len(s.ams)
+	groupID %= groupSize
+	for j := 0; j < groupSize; groupID, j = (groupID+1)%groupSize, j+1 {
```

Comment:
why not foreach `s.ams` or `s.kgs`?

Reply:
because "foreach s.ams and s.kgs" brings O(n^2) time complexity to adjust all primary locations based on their priorities. This function is called every 10 seconds, and each time it only finds the next primary which needs to adjust location then exit the loop. For better efficiency, when scanning the array each time, it starts from the position left by the last scan.